ethcore/src/snapshot/service.rs

```diff
@@ -406,55 +416,133 @@ impl Service {
 	/// Initialize the restoration synchronously.
 	/// The recover flag indicates whether to recover the restored snapshot.
 	pub fn init_restore(&self, manifest: ManifestData, recover: bool) -> Result<(), Error> {
-		let rest_dir = self.restoration_dir();
-
-		let mut res = self.restoration.lock();
+		self.ready.store(false, Ordering::SeqCst);
 
-		self.state_chunks.store(0, Ordering::SeqCst);
-		self.block_chunks.store(0, Ordering::SeqCst);
-
-		// tear down existing restoration.
-		*res = None;
+		let rest_dir = self.restoration_dir();
+		let rest_db = self.restoration_db();
+		let recovery_temp = self.temp_recovery_dir();
+		let prev_chunks = self.prev_chunks_dir();
 
 		// delete and restore the restoration dir.
-		if let Err(e) = fs::remove_dir_all(&rest_dir) {
+		if let Err(e) = fs::remove_dir_all(&prev_chunks) {
 			match e.kind() {
 				ErrorKind::NotFound => {},
 				_ => return Err(e.into()),
 			}
 		}
 
-		fs::create_dir_all(&rest_dir)?;
+		// Move the previous recovery temp directory
+		// to `prev_chunks` to be able to restart restoring
+		// with previously downloaded blocks
+		// This step is optional, so don't fail on error
+		fs::rename(&recovery_temp, &prev_chunks).ok();
```

Comment:
IMHO this should be moved under `restoration.lock()`, previously it was ok to remove and create the same directory multiple times, now we are doing a rename which should probably be done only once.

Reply:
(this is done now)